src/features/cart/ShoppingCartTest.ts

```diff
@@ -0,0 +1,185 @@
+export {}
+interface Item{
+    id:number,
+    title:string,
+    price:number,
+    quantity:number,
+    discount:string[]
+};
+
+
+interface DiscountCode{
+    code:string,
+    type: "cart" | "item",
+    value:number,
+    itemId?: number;
+
+};
+interface Cart{
+    items:Item[],
+    discount:string[]
+}
+
+export const discountCodes:Array<DiscountCode>=[
+    {
+        code:'NewYear2022',
+        type:"cart",
+        value:20,
+    },
+    {
+        code:'summer2022',
+        type:"cart",
+        value:10,
+    },
+    {
+        code:'fullFree',
+        type:"item",
+        value:25,
+        itemId:2
+        
+    }
+
+]
+export const cart:Cart={
+    items:[],
+    discount:[]
+}
+
+//Add to Product Card 
+export const addToCart=(item:Item):void=>{
+    const isItem=cart.items.find((product)=>product.id===item.id)
+    if(isItem){
+        isItem.quantity++;
+        // item.quantity++
+    }else{
+        cart.items.push(item)
+        console.log("New Item Add" ,cart.items.find((product)=>product.id==item.id),)
+    }
+};
+export const removeToCart=(id:number):void=>{
+    const isItem=cart.items.find((product)=>product.id==id && product.quantity>=1)
+    if(isItem.quantity>1){
+        isItem.quantity--,
+        console.log("One Item Remove ",isItem)
+
+    }
+    else if(isItem.quantity=1){
+        cart.items=cart.items.filter((item)=>item.id!==isItem.id)
+        // cart.items.filter((item)=>item.id !==isItem.id)
+        console.log("No Item From this Cart")
+
+    }
+    else{
+        console.log("No item form this Cart",id)
+    }
+}
+export const addDiscountCode=(discountCode:string):void=>{
+    //Find the Code 
+    const checkCode=discountCodes.find((item)=>item.code==discountCode)
+    if(checkCode){
+        if(checkCode.type="cart"){
+            //all Ready Use this Code Or not 
+            const isApplied=cart.discount.includes(checkCode.code)
+            if(!isApplied){
+                cart.discount.push(checkCode.code)
+            }
+            else{
+                console.log("This Code allready Exits in the Cart",isApplied)
+                return
+            }
+        }
+        else{
+            const product=cart.items[cart.items.findIndex((item)=>item.id==checkCode.itemId)]
+            //All ready Use or not
+            const isApplid=product.discount.includes(checkCode.code)
+            if(!isApplid){
+                product.discount.push(checkCode.code)
+                console.log("Descount Code is item",product)
+
+            }else{
+                console.log("All Ready Use this code before")
+                return
+            }
+
+        }
+
+    }
+    else{
+        console.log("There Is No Code Here ",discountCode)
+    }
+
+}
+//Get Single Card Price 
```

Comment:
Thank you for adding comments above the function describing what it is doing. There is a formal format to add comments to functions as follows.

```
/**
function description text goes here
@param a - description for parameter a
@param b - description for parameter b
@returns description of what it returns
**/
function add(a: number, b: number): number {

}
```

Checkout this for reference.
https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html

It would be great if you could go through this documentation and present to the team to enrich everyone.

Reply:
Typically we must add comments to functions that are exported from a module (file). For non-exported function it is not mandatory. Use your discretion for those. If it  is a complex function, it is a good practice to add documentation comment on top of those non-exported functions.